File: app/src/main/java/com/androiddev/social/timeline/data/TimeUtils.java
package com.androiddev.social.timeline.data;

import static java.util.concurrent.TimeUnit.DAYS;
import static java.util.concurrent.TimeUnit.HOURS;
import static java.util.concurrent.TimeUnit.MINUTES;
import static java.util.concurrent.TimeUnit.SECONDS;

import java.util.Arrays;
import java.util.List;

/**
 * Utilities for dealing with dates and times
 */
public class TimeUtils {

    public static final List<Long> times = Arrays.asList(
        DAYS.toMillis(365),
        DAYS.toMillis(30),
        DAYS.toMillis(7),
        DAYS.toMillis(1),
        HOURS.toMillis(1),
        MINUTES.toMillis(1),
        SECONDS.toMillis(1)
    );

    public static final List<String> timesString = Arrays.asList(
        "yr", "mo", "wk", "day", "hr", "min", "sec"
    );

    /**
     * Get relative time ago for date
     *
     * NOTE:
     *  if (duration > WEEK_IN_MILLIS) getRelativeTimeSpanString prints the date.
     *
     * ALT:
     *  return getRelativeTimeSpanString(date, now, SECOND_IN_MILLIS, FORMAT_ABBREV_RELATIVE);
     *
     * @param date String.valueOf(TimeUtils.getRelativeTime(1000L * Date/Time in Millis)
     * @return relative time
     */
    public static CharSequence getRelativeTime(final long date) {
        return toDuration( Math.abs(System.currentTimeMillis() - date) );
    }

    private static String toDuration(long duration) {
        StringBuilder sb = new StringBuilder();
        for(int i=0;i< times.size(); i++) {
            Long current = times.get(i);
            long temp = duration / current;
            if (temp > 0) {
                sb.append(temp)
                  .append(" ")
                  .append(timesString.get(i))
                  .append(temp > 1 ? "s" : "")
                  .append(" ago");
                break;
            }
        }
        return sb.toString().isEmpty() ? "now" : sb.toString();
    }
}

File: app/src/main/java/com/androiddev/social/timeline/data/TimeUtils.kt
package com.androiddev.social.timeline.data

import java.util.Calendar
import java.util.Date

object FuzzyDateTimeFormatter {
    private const val SECONDS = 1
    private const val MINUTES = 60 * SECONDS
    private const val HOURS = 60 * MINUTES
    private const val DAYS = 24 * HOURS
    private const val WEEKS = 7 * DAYS
    private const val MONTHS = 4 * WEEKS
    private const val YEARS = 12 * MONTHS

    /**
     * Returns a properly formatted fuzzy string representing time ago
     * @param context   Context
     * @param date      Absolute date of the event
     * @return          Formatted string
     */
    fun Date.getTimeAgo(): String {
        val date = this
        val beforeSeconds = (date.time / 1000).toInt()
        val nowSeconds = (Calendar.getInstance().timeInMillis / 1000).toInt()
        val timeDifference = nowSeconds - beforeSeconds
        require(timeDifference >= 0) { "Date must be in the past!" }
        return if (timeDifference < 15 * SECONDS) {
            "just now"
        } else if (timeDifference < MINUTES) {
            "$timeDifference second ago"
        } else if (timeDifference < HOURS) {
            "$timeDifference hours ago"

        } else if (timeDifference < DAYS) {
            "$timeDifference days ago"

        } else if (timeDifference < WEEKS) {
            "$timeDifference weeks ago"

        } else if (timeDifference < MONTHS) {
            "$timeDifference months ago"

        } else if (timeDifference < YEARS) {
            "$timeDifference years ago"

        } else {
            "$timeDifference years ago"
        }
    }
}